File: kernel/kernel.lua
local Vfs = require("vfs.vfs")
local ProcessManager = require("processes.manager")
local RamFS = require("drivers.agnostic.ramfs")
local SerialFS = require("drivers.agnostic.serialfs")
local InodeModeFlags = require("vfs.inode.modeflags")
local Process = require("processes.process")

---@class Kernel
---@field arch Arch
---@field vfs Vfs
---@field procman ProcessManager
local Kernel = {}

function Kernel:new(arch)
	local kernel = {
		arch = arch,
		vfs = Vfs:create(),
		procman = ProcessManager:new(),
	}

	setmetatable(kernel, self)
	self.__index = self

	return kernel
end

function Kernel:start()
	-- Mount agnostic filesystems and devices
	local rootfs = SerialFS:new(self.arch, "rootfs.dat")
	self.vfs:mount(rootfs, "/")

	-- Ensure the /dev mountpoint directory exists in rootfs before mounting devfs.
	-- SerialFS restores from disk, so it may already be present.
	if not rootfs:get_inode("/dev") then
		rootfs:create_file(self.vfs.root_mount.root_inode, "dev", InodeModeFlags.TYPE_DIR)
	end

	local devfs = RamFS:new(self.arch)
	self.vfs:mount(devfs, "/dev")
	-- Mount devices
	self.arch:init(self)

	local init_process = Process:new(self, "/bin/init.lua")
	self.procman:add_process(init_process)
	while true do
		self.arch:step()
		self.procman:step()
		local init = self.procman.processes[1]
		if not init or init.dead then
			error("kernel panic: init (PID 1) died with exit code " .. tostring(init and init.exit_code))
		end
	end
end

return Kernel

File: kernel/kernel.ts
import Process from "processes.process";
import { Arch } from "./arch/arch";
import { RamFS } from "./drivers/agnostic/ramfs";
import { SerialFS } from "./drivers/agnostic/serialfs";
import { ProcessManager } from "./processes/manager";
import { Vfs } from "./vfs/vfs";

export class Kernel {
	arch: Arch;
	vfs: Vfs = new Vfs();
	procman: ProcessManager = new ProcessManager();

	constructor(arch: Arch) {
		this.arch = arch;
	}

	start() {
		let rootfs = new SerialFS(this.arch, "rootfs.dat");
		this.vfs.mount(rootfs, "/");

		if (rootfs.get_inode("/dev") === undefined) {
			rootfs.create_file(this.vfs.root_mount?.root_inode!, "dev", InodeModeFlags.TYPE_DIR)
		}

		let devfs = new RamFS(this.arch);
		this.vfs.mount(devfs, "/dev");
		this.arch.init(this);

		let init_process = new Process(this, "/bin/init.lua");
		this.procman.add_process(init_process);

		while (true) {
			this.arch.step();
			this.procman.step();
			let init = this.procman.get_process(1);
			if (init === null || init.dead) {
				throw "kernel panic: init (PID 1) died with exit code " + tostring(init && init.exit_code);
			}
		}
	}
}
